### PR TITLE
feat(sliding sync): remove `response_handling_lock` and extend the position's lock responsibilities

### PR DESCRIPTION
**Pull request by bnjbvr**
In the previous situation, we had two locks with similar responsibilities, the `response_handling_lock` and the `position` lock. The latter *almost* covered the former's critical zone, albeit for a single function call, which left room for a deadlock situation (latter taken, then former, then latter).

This removes the former, and extends the critical zone of the latter up to the end of the response handling, removing the possibility of the deadlock entirely.

Fixes https://github.com/matrix-org/matrix-rust-sdk/issues/2426.

### CODE REVIEW

**Review by Hywan — COMMENTED** [review]
Looks good so far!

File: crates/matrix-sdk/src/sliding_sync/mod.rs [hunk 1]
```diff
@@ -1945,4 +1927,43 @@ mod tests {
...
+        let sliding_sync = client
+            .sliding_sync("test")?
+            .with_to_device_extension(
+                assign!(v4::ToDeviceConfig::default(), { enabled: Some(true)}),
+            )
+            .with_e2ee_extension(assign!(v4::E2EEConfig::default(), { enabled: Some(true)}))
+            .build()
+            .await?;
+
+        // Spawn two requests in parallel. Before #2430, this lead to a deadlock and the
```
Comment (Hywan) [inline crates/matrix-sdk/src/sliding_sync/mod.rs:1959]:
```suggestion
        // Spawn two requests in parallel. Before #2430, this leads to a deadlock and the
```

**Review by Hywan — APPROVED** [review]
Awesome!

File: crates/matrix-sdk/src/sliding_sync/mod.rs (hunk 1 above)
Comment (bnjbvr) [reply crates/matrix-sdk/src/sliding_sync/mod.rs:1959]: I meant this as the past form for "lead" (which is "lead"), so I'll leave it to that :p

**Pull request merged**